tests/testAirPressure.cpp

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "testAirPressure.h"
+#include "core/divesite.h"
+#include "core/divelist.h"
+#include "core/file.h"
+#include <QString>
+#include <core/qthelper.h>
+
+void TestAirPressure::initTestCase()
+{
+	/* we need to manually tell that the resource exists, because we are using it as library. */
+	Q_INIT_RESOURCE(subsurface);
+}
+
+void TestAirPressure::get_dives()
+{
+	struct dive *dive;
+	verbose = 1;
+
+	QCOMPARE(parse_file(SUBSURFACE_TEST_DATA "/dives/TestAtmPress.xml", &dive_table, &trip_table, &dive_site_table), 0);
+	dive = get_dive(0);
+	dive->selected = true;
+	QVERIFY(dive != NULL);
+}
+
+void TestAirPressure::testReadAirPressure()
+{
+	struct dive *dive;
+	dive = get_dive(0);
+	dive->selected = true;
+	QVERIFY(dive != NULL);
+	QCOMPARE(QString::number(1012), QString::number(dive->surface_pressure.mbar));
+	dive = get_dive(1);
+	dive->selected = true;
+	QVERIFY(dive != NULL);
+	if (dive->surface_pressure.mbar)
+	QCOMPARE(QString::number(991), QString::number(dive->surface_pressure.mbar));
+}
+
+void TestAirPressure::testWriteReadBackAirPressure()
+{
+	struct dive *dive;
```

Comment:
just like above, you have the ordering of dereferencing and non-null tests as well as the odd conversion into string before using `QCOMPARE()`. Oddly, you are happy to use `QCOMPARE()` on numbers when you check the return value of functions... :-)

Reply:
Do not feel bad at all. This is what is required to get the code up to standard. Your comments are appreciated. Maybe we should talk about the flag. It is an extra variable in the dive structure and I would like to limit adding extra variables to that structure. I do not want to run ahead of things and try to limit the discussion to the airpressure, but it is possible that other similar coding might appear in the near future. I am thinking of salinity in the case of manually-entered dives. I need to get your opinion about adding more booleans if other similar cases in coding do appear in future. My idea was basically to turn the userflag into a byte and do some bit-setting and bit-reading using | and & operators and thus avoid having to add additional variables to the dive structure. If you are not too worried about adding a couple of additional booleans if the need may arise in future, then probably using a byte is not necessary. I am affected by my programming past where every byte counted. What is your feeling?

I am working on your other requests.

Reply:
> Maybe we should talk about the flag. It is an extra variable in the dive structure and I would like to limit adding extra variables to that structure.

I'm not saying that the flag is wrong. I'd just like to note that for air temperature, there is a different mechanism: There is an `un_fixup_airtemp` function that returns 0 if the airtemp corresponds to the one supplied by the dive computer.

The analog would look something like this:
```
diff --git a/core/dive.c b/core/dive.c
index 114ca31db..55498ebdd 100644
--- a/core/dive.c
+++ b/core/dive.c
@@ -1317,9 +1317,10 @@ static struct event *find_previous_event(struct divecomputer *dc, struct event *
 	return previous;
 }
 
-static void fixup_surface_pressure(struct dive *dive)
+static pressure_t calculate_surface_pressure(const struct dive *dive)
 {
-	struct divecomputer *dc;
+	pressure_t res;
+	const struct divecomputer *dc;
 	int sum = 0, nr = 0;
 
 	for_each_dc (dive, dc) {
@@ -1328,8 +1329,13 @@ static void fixup_surface_pressure(struct dive *dive)
 			nr++;
 		}
 	}
-	if (nr)
-		dive->surface_pressure.mbar = (sum + nr / 2) / nr;
+	res.mbar = nr ? (sum + nr / 2) / nr : 0;
+	return res;
+}
+
+static void fixup_surface_pressure(struct dive *dive)
+{
+	dive->surface_pressure = calculate_surface_pressure(dive);
 }
 
 static void fixup_water_salinity(struct dive *dive)
@@ -1424,6 +1430,17 @@ static void fixup_airtemp(struct dive *dive)
 		dive->airtemp.mkelvin = dc_airtemp(&dive->dc);
 }
 
+/* if the surface pressure in the dive data is redundant to the calculated
+ * value (i.e., it was added by running fixup on the dive) return 0,
+ * otherwise return the air temperature given in the dive */
+pressure_t un_fixup_surface_pressure(const struct dive *d)
+{
+	pressure_t res = d->surface_pressure;
+	if (res.mbar && res.mbar == calculate_surface_pressure(d).mbar)
+		res.mbar = 0;
+	return res;
+}
+
 /* if the air temperature in the dive data is redundant to the one in its
  * first divecomputer (i.e., it was added by running fixup on the dive)
  * return 0, otherwise return the air temperature given in the dive */
diff --git a/core/dive.h b/core/dive.h
index 7a6cf4099..f1f2cc36b 100644
--- a/core/dive.h
+++ b/core/dive.h
@@ -536,6 +536,7 @@ extern void sort_dive_table(struct dive_table *table);
 extern void sort_trip_table(struct trip_table *table);
 extern struct dive *fixup_dive(struct dive *dive);
 extern void fixup_dc_duration(struct divecomputer *dc);
+extern pressure_t un_fixup_surface_pressure(const struct dive *d);
 extern int dive_getUniqID();
 extern unsigned int dc_airtemp(const struct divecomputer *dc);
 extern unsigned int dc_watertemp(const struct divecomputer *dc);
diff --git a/core/save-git.c b/core/save-git.c
index ffbeca1d0..c6cb0c969 100644
--- a/core/save-git.c
+++ b/core/save-git.c
@@ -419,11 +419,12 @@ static void save_dc(struct membuffer *b, struct dive *dive, struct divecomputer
  */
 static void create_dive_buffer(struct dive *dive, struct membuffer *b)
 {
+	pressure_t surface_pressure = un_fixup_surface_pressure(dive);
 	if (dive->dc.duration.seconds > 0)
 		put_format(b, "duration %u:%02u min\n", FRACTION(dive->dc.duration.seconds, 60));
 	SAVE("rating", rating);
 	SAVE("visibility", visibility);
-	if (dive->surface_pressure.mbar)
+	if (surface_pressure.mbar)
 		SAVE("airpressure", surface_pressure.mbar);
 	cond_put_format(dive->notrip, b, "notrip\n");
 	save_tags(b, dive->tag_list);
diff --git a/core/save-xml.c b/core/save-xml.c
index 027c6f4f2..8b5e3a7e6 100644
--- a/core/save-xml.c
+++ b/core/save-xml.c
@@ -473,6 +473,7 @@ static void save_picture(struct membuffer *b, struct picture *pic)
 void save_one_dive_to_mb(struct membuffer *b, struct dive *dive, bool anonymize)
 {
 	struct divecomputer *dc;
+	pressure_t surface_pressure = un_fixup_surface_pressure(dive);
 
 	put_string(b, "<dive");
 	if (dive->number)
@@ -488,8 +489,8 @@ void save_one_dive_to_mb(struct membuffer *b, struct dive *dive, bool anonymize)
 		put_format(b, " divesiteid='%8x'", dive->dive_site->uuid);
 	}
 	show_date(b, dive->when);
-	if (dive->surface_pressure.mbar)
-		put_format(b, " airpressure='%d'", dive->surface_pressure.mbar);
+	if (surface_pressure.mbar)
+		put_format(b, " airpressure='%d'", surface_pressure.mbar);
 	if (dive->dc.duration.seconds > 0)
 		put_format(b, " duration='%u:%02u min'>\n",
 			   FRACTION(dive->dc.duration.seconds, 60));
```

But let me reiterate that I'm *not* saying that this is preferred over a flag. I only want to mention that as an alternative...

Reply:
I really like this... better than carrying more flags. And of course more consistent